core/drivers/stm32_bsec.c

```diff
@@ -0,0 +1,599 @@
+	while (!timeout_elapsed(timeout_ref))
+		if (!(bsec_status() & BSEC_MODE_BUSY_MASK))
+			break;
+
+	if (bsec_status() & (BSEC_MODE_BUSY_MASK | BSEC_MODE_PROGFAIL_MASK))
+		result = TEE_ERROR_GENERIC;
+	else
+		result = check_no_error(otp_id);
+
+	power_down_safmem();
+
+	bsec_unlock(exceptions);
+
+	return result;
+}
+
+TEE_Result stm32_bsec_permanent_lock_otp(uint32_t otp_id)
+{
+	TEE_Result result = 0;
+	uint32_t data = 0;
+	uint32_t addr = 0;
+	uint32_t exceptions = 0;
+	vaddr_t base = bsec_base();
+	uint64_t timeout_ref;
+
+	if (otp_id > otp_max_id())
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (otp_id < bsec_dev.upper_base) {
+		addr = otp_id >> ADDR_LOWER_OTP_PERLOCK_SHIFT;
+		data = DATA_LOWER_OTP_PERLOCK_BIT <<
+		       ((otp_id & DATA_LOWER_OTP_PERLOCK_MASK) << 1U);
+	} else {
+		addr = (otp_id >> ADDR_UPPER_OTP_PERLOCK_SHIFT) + 2U;
+		data = DATA_UPPER_OTP_PERLOCK_BIT <<
+		       (otp_id & DATA_UPPER_OTP_PERLOCK_MASK);
+	}
+
+	exceptions = bsec_lock();
+
+	result = power_up_safmem();
+	if (result)
+		return result;
+
+	io_write32(base + BSEC_OTP_WRDATA_OFF, data);
+	io_write32(base + BSEC_OTP_CTRL_OFF, addr | BSEC_WRITE | BSEC_LOCK);
+
+	timeout_ref = timeout_init_us(BSEC_TIMEOUT_US);
+	while (!timeout_elapsed(timeout_ref))
+		if (!(bsec_status() & BSEC_MODE_BUSY_MASK))
+			break;
+
+	if (bsec_status() & (BSEC_MODE_BUSY_MASK | BSEC_MODE_PROGFAIL_MASK))
+		result = TEE_ERROR_BAD_PARAMETERS;
+	else
+		result = check_no_error(otp_id);
+
+	power_down_safmem();
+
+	bsec_unlock(exceptions);
+
+	return result;
+}
+
+TEE_Result stm32_bsec_write_debug_conf(uint32_t value)
+{
+	TEE_Result result = TEE_ERROR_GENERIC;
+	uint32_t masked_val = value & BSEC_DEN_ALL_MSK;
+	unsigned int exceptions = 0;
+
+	exceptions = bsec_lock();
+
+	io_write32(bsec_base() + BSEC_DEN_OFF, value);
+
+	if ((io_read32(bsec_base() + BSEC_DEN_OFF) ^ masked_val) == 0U)
+		result = TEE_SUCCESS;
+
+	bsec_unlock(exceptions);
+
+	return result;
+}
+
+uint32_t stm32_bsec_read_debug_conf(void)
+{
+	return io_read32(bsec_base() + BSEC_DEN_OFF);
+}
+
+bool stm32_bsec_write_sr_lock(uint32_t otp_id, uint32_t value)
+{
+	uint32_t bank = otp_bank_offset(otp_id);
+	uint32_t bank_value = 0;
+	uint32_t otp_mask = BIT(otp_id & BSEC_OTP_MASK);
+	uint32_t exceptions = 0;
+
+	if (!value)
+		return false;
+
+	exceptions = bsec_lock();
+
+	bank_value = io_read32(bsec_base() + BSEC_SRLOCK_OFF + bank);
+
+	if ((bank_value & otp_mask) != value) {
+		/*
+		 * We can write 0 in all other OTP
+		 * if the lock is activated in one of other OTP.
+		 * Write 0 has no effect.
+		 */
+		io_write32(bsec_base() + BSEC_SRLOCK_OFF + bank,
+			   bank_value | otp_mask);
+	}
+
+	bsec_unlock(exceptions);
+
+	return true;
+}
+
+bool stm32_bsec_read_sr_lock(uint32_t otp_id)
+{
+	uint32_t bank = otp_bank_offset(otp_id);
+	uint32_t otp_mask = BIT(otp_id & BSEC_OTP_MASK);
+	uint32_t bank_value = io_read32(bsec_base() + BSEC_SRLOCK_OFF + bank);
+
+	return (bank_value & otp_mask) != 0U;
```

Comment:
This is the same thing as `return bank_value & otp_mask;`, given that the function returns a `bool`